[dask-on-ray] Add better Dask-on-Ray example, and detail custom shuffle optimization. (#13950)
ray-project/ray: doc/source/dask-on-ray.rst

Dask on Ray
===========

.. _dask-on-ray:

`Dask <https://dask.org/>`__ is a Python parallel computing library geared towards scaling analytics and
scientific computing workloads. It provides `big data collections
<https://docs.dask.org/en/latest/user-interfaces.html>`__ that mimic the APIs of
the familiar `NumPy <https://numpy.org/>`__ and `Pandas <https://pandas.pydata.org/>`__ libraries, 
allowing those abstractions to represent
larger-than-memory data and/or allowing operations on that data to be run on a multi-machine cluster, 
while also providing automatic data parallelism, smart scheduling,
and optimized operations. Operations on these collections create a task graph, which is
executed by a scheduler.

> Dask-on-Ray is a scheduler for Dask, allowing you to build data
> analyses using Dask's collections and then execute
> the underlying tasks on a Ray cluster. We use Dask's scheduler API, which allows you to
> specify any callable as the scheduler that you would like Dask to use to execute your

Reviewer: ```suggestion
Ray provides a scheduler for Dask (`dask_on_ray`) which allows you to build data
analyses using Dask's collections and execute
the underlying tasks on a Ray cluster. 

`dask_on_ray` uses Dask's scheduler API, which allows you to
specify any callable as the scheduler that you would like Dask to use to execute your
```

Reviewer: (nit: maintain 3rd or 2nd person)

.. note::

>   Note that Ray does not yet support object spilling, and hence cannot

Reviewer 2: Also, it'll be great how we achieved this (mention we used Dask's plugin APIs to implement this).

>   Note that Ray does not yet support object spilling, and hence cannot

Author: I don't think that we will ever want to implement Dask Distributed's scheduler protocol and/or the client API, at least not in its current form. I think that it's more likely that we would offer our own client that offers an API that better fits the task/cluster state introspection and management capabilities and APIs that Ray exposes.

>   process datasets larger than cluster memory. This feature is currently being developed.

Author: But object spilling won't be turned on by default until the 1.3 release, right? Should we wait to call out that feature until it's on by default in the most recent release, or should we call it out and include instructions for turning it on?

>   process datasets larger than cluster memory. This feature is currently being developed.

Reviewer 2: I think it is fine because we already have a doc in 1.2.0 about how to turn on the object spilling. The feature is probably not as stable as 1.3.0, but we can mention the feature is under active development and report issues if they face any.

   import ray
   from ray.util.dask import ray_dask_get
   import dask.array as da
   import dask.dataframe as dd
   import numpy as np
   import pandas as pd
   import time

   # Start Ray.
   # Tip: If you're connecting to an existing cluster, use ray.init(address="auto").
   ray.init()

   d_arr = da.from_array(np.random.randint(0, 1000, size=(256, 256)))

   # The Dask scheduler submits the underlying task graph to Ray.
   d_arr.mean().compute(scheduler=ray_dask_get)

   # Set the scheduler to ray_dask_get in your config so you don't have to specify it on
   # each compute call.
   dask.config.set(scheduler=ray_dask_get)

   df = dd.from_pandas(pd.DataFrame(np.random.randint(0, 100, size=(1024, 2)), columns=["age", "grade"]))

>    df.groupby(["age"]).mean().compute()

Reviewer 2: I think we should write more examples in a different section, but we can probably do that later.

Reviewer 2: Also, can you write a note under this section that every Dask API should be available, but if users find any issue, create an issue to our github issue page?

Why use Dask on Ray?

   1. If you'd like to create data analyses using the familiar NumPy and Pandas APIs provided by Dask and execute them on a fast, fault-tolerant distributed task execution system geared towards production, like Ray.
   2. If you'd like to use Dask and Ray libraries in the same application without having two different clusters.

>    3. To take advantage of Ray-specific features such as the

Reviewer 2: Please also mention the mature actor implementation. Also, let's put this as the first reason (since I believe it is the strongest benefit).

Author: That's a good motivation for using Ray in general, but Dask-on-Ray doesn't use actors, so I think that calling out such an orthogonal feature doesn't really fit here. We can point them to a more general "why use Ray" docs page that contains a more comprehensive listing of Ray's features, but I think that we should be calling out Ray features that Dask workloads will benefit from here.

Reviewer 2: Hmm yeah a good point about actors. But shared memory is definitely a useful feature. Should we only mention shared memory in this case?

Author: Shared memory is mentioned, or are you saying that you think that shared memory should go first, before mentioning the cluster launcher?

Reviewer 2: Yeah I think the first reason is the least compelling reason compared to other 2. Maybe we can change the order? 2 -> 3 -> 1 or 3 -> 2 -> 1

      :ref:`cluster launcher <ref-automatic-cluster>` and
      :ref:`shared-memory store <memory>`.

> Note that for execution on a Ray cluster, you should *not* use the

Reviewer 2: Put this under 

```
.. note::
```
section

`Dask.distributed <https://distributed.dask.org/en/latest/quickstart.html>`__

> client; simply use plain Dask and its collections, and pass ``ray_dask_get``

Reviewer 2: Please explain this is "currently" not supported. (We can probably starting supporting it once ray client becomes more stable)

Author: I don't think that we will ever want to implement Dask Distributed's scheduler protocol and/or the client API, at least not in its current form. I think that it's more likely that we would offer our own client that offers an API that better fits the task/cluster state introspection and management capabilities and APIs that Ray exposes.

to ``.compute()`` calls or set the scheduler in one of the other ways detailed `here <https://docs.dask.org/en/latest/scheduling.html#configuration>`__. Follow the instructions for
:ref:`using Ray on a cluster <using-ray-on-a-cluster>` to modify the
``ray.init()`` call.

Dask-on-Ray is an ongoing project and is not expected to achieve the same performance as using Ray directly.

================================================
Custom optimization for Dask DataFrame shuffling
================================================

We have also created a custom Dask DataFrame optimizer that leverages Ray's ability to

> execute multiple-return tasks in order to speed up shuffling by as much as 10x on Ray.

Author: I got a 10x speedup on my local machine for a smaller shuffle, but we can limit it to speedups that we got on our more official benchmarks if you'd like.

> execute multiple-return tasks in order to speed up shuffling by as much as 10x on Ray.

Reviewer 2: Maybe 4~10X?

> execute multiple-return tasks in order to speed up shuffling by as much as 10x on Ray.

Author: I'll keep it to the official benchmarks and say 4x.

Simply set the `dataframe_optimize` configuration option to our optimizer function, similar to how you specify the Dask-on-Ray scheduler:

.. code-block:: python

   import ray
   from ray.util.dask import ray_dask_get, dataframe_optimize
   import dask.dataframe as dd
   import numpy as np
   import pandas as pd
   import time

   # Start Ray.
   # Tip: If you're connecting to an existing cluster, use ray.init(address="auto").
   ray.init()

   # Set the scheduler to ray_dask_get, and set the Dask DataFrame optimizer to our
   # custom optimization function, this time using the config setter as a context manager.

>    with dask.config.set(scheduler=ray_dask_get, dataframe_optimize=dataframe_optimize):

Reviewer 3: Can we enable these by default on ray.init()?

Author: We can, although I think that would be less good for a few reasons:

1. Having the core `ray.init()` path reach out and set Dask configs seems like a poor separation of concerns, and may be odd for those using both Dask and Ray but _not_ trying to use Dask-_on_-Ray.
2. If the user wants to compare the two, they wouldn't want that config set globally, and might find that confusing.

I personally think that making the user explicitly set a config to enable Dask-on-Ray is the better way to go. If you think that this configuration is getting ugly (and could get uglier if we add more Ray-specific optimization passes), we could provide a helper like `enable_dask_on_ray()` that hides the complexity of manually setting these Dask configs.

Reviewer 3: Makes sense. I think adding something like `ray.util.enable_dask_integration()` would be very helpful for users.

Author: Agreed.

Reviewer 4: This should be pretty quick, right? @clarkzinzow, do you want to do that in this PR?

Author: That should be very quick, yes, but I'll keep this PR focused on docs and open another PR that adds that helper.

>        df = dd.from_pandas(pd.DataFrame(np.random.randint(0, 100, size=(10000, 2)), columns=["age", "grade"]))

Reviewer 2: We should mention that max_branch should be large in this case.

   df = dd.from_pandas(pd.DataFrame(np.random.randint(0, 100, size=(1024, 2)), columns=["age", "grade"]))
   df.groupby(["age"]).mean().compute()

.. note::
  For execution on a Ray cluster, you should *not* use the
  `Dask.distributed <https://distributed.dask.org/en/latest/quickstart.html>`__
  client; simply use plain Dask and its collections, and pass ``ray_dask_get``
  to ``.compute()`` calls or set the scheduler in one of the other ways detailed `here <https://docs.dask.org/en/latest/scheduling.html#configuration>`__. Follow the instructions for
  :ref:`using Ray on a cluster <using-ray-on-a-cluster>` to modify the
  ``ray.init()`` call.

Why use Dask on Ray?

   1. If you'd like to create data analyses using the familiar NumPy and Pandas APIs provided by Dask and execute them on a fast, fault-tolerant distributed task execution system geared towards production, like Ray.
   2. If you'd like to use Dask and Ray libraries in the same application without having two different clusters.
   3. To take advantage of Ray-specific features such as the
      :ref:`cluster launcher <ref-automatic-cluster>` and
      :ref:`shared-memory store <memory>`.

> Dask-on-Ray is an ongoing project and is not expected to achieve the same performance as using Ray directly. All `Dask abstractions <https://docs.dask.org/en/latest/user-interfaces.html>`__ should run seamlessly on top of Ray using this scheduler, so if you find that one of these abstractions doesn't run on Ray, please open an issue on the `Ray GitHub repo <https://github.com/ray-project/ray>`__.

Reviewer 4: ```suggestion
Dask-on-Ray is an ongoing project and is not expected to achieve the same performance as using Ray directly. All `Dask abstractions <https://docs.dask.org/en/latest/user-interfaces.html>`__ should run seamlessly on top of Ray using this scheduler, so if you find that one of these abstractions doesn't run on Ray, please `open an issue <https://github.com/ray-project/ray/issues/new/choose>`__.
```

================================================
Out-of-Core Data Processing

> ================================================

Reviewer 2: Can you align the match the length of the string and `=`? 

```
=====================
Out-of-Core Data Processing

```

Processing datasets larger than cluster memory is supported via Ray's `object spilling

> <https://docs.ray.io/en/master/memory-management.html#object-spilling>`__: if

Reviewer 2: We shouldn't tag master here directly. There's a way to make a link natively. (Use this as an example; https://github.com/ray-project/ray/blob/master/doc/examples/placement-group.rst)

   import ray
   from ray.util.dask import ray_dask_get
   import dask.array as da
   import dask.dataframe as dd
   import numpy as np
   import pandas as pd
   import time

   # Start Ray.
   # Tip: If you're connecting to an existing cluster, use ray.init(address="auto").
   ray.init()

   d_arr = da.from_array(np.random.randint(0, 1000, size=(256, 256)))

   # The Dask scheduler submits the underlying task graph to Ray.
   d_arr.mean().compute(scheduler=ray_dask_get)

   # Set the scheduler to ray_dask_get in your config so you don't have to specify it on
   # each compute call.
   dask.config.set(scheduler=ray_dask_get)

>    df = dd.from_pandas(pd.DataFrame(np.random.randint(0, 100, size=(1024, 2)), columns=["age", "grade"]))

Reviewer: this is a bit hard to read on the docs; try 80 column width?

   df.groupby(["age"]).mean().compute()

.. note::
  For execution on a Ray cluster, you should *not* use the
  `Dask.distributed <https://distributed.dask.org/en/latest/quickstart.html>`__
  client; simply use plain Dask and its collections, and pass ``ray_dask_get``
  to ``.compute()`` calls or set the scheduler in one of the other ways detailed `here <https://docs.dask.org/en/latest/scheduling.html#configuration>`__. Follow the instructions for
  :ref:`using Ray on a cluster <using-ray-on-a-cluster>` to modify the
  ``ray.init()`` call.

Why use Dask on Ray?

>    1. To take advantage of Ray-specific features such as the

Reviewer: spacing turns out weird
```suggestion
1. To take advantage of Ray-specific features such as the
```

      :ref:`cluster launcher <ref-automatic-cluster>` and
      :ref:`shared-memory store <memory>`.

>    2. If you'd like to use Dask and Ray libraries in the same application without having two different clusters.
>    3. If you'd like to create data analyses using the familiar NumPy and Pandas APIs provided by Dask and execute them on a fast, fault-tolerant distributed task execution system geared towards production, like Ray.

Reviewer: the spacing turns out weird

```suggestion
2. If you'd like to use Dask and Ray libraries in the same application without having two different clusters.
3. If you'd like to create data analyses using the familiar NumPy and Pandas APIs provided by Dask and execute them on a fast, fault-tolerant distributed task execution system geared towards production, like Ray.
```

Dask-on-Ray is an ongoing project and is not expected to achieve the same performance as using Ray directly. All `Dask abstractions <https://docs.dask.org/en/latest/user-interfaces.html>`__ should run seamlessly on top of Ray using this scheduler, so if you find that one of these abstractions doesn't run on Ray, please `open an issue <https://github.com/ray-project/ray/issues/new/choose>`__.

Out-of-Core Data Processing
---------------------------

.. _dask-on-ray-out-of-core:

Processing datasets larger than cluster memory is supported via Ray's :ref:`object spilling <object-spilling>`: if
the in-memory object store is full, objects will be spilled to external storage (local disk by
default). This feature is available but off by default in Ray 1.2, and is on by default
in Ray 1.3+. Please see your Ray version's object spilling documentation for steps to enable and/or configure
object spilling.

Custom optimization for Dask DataFrame shuffling
------------------------------------------------

.. _dask-on-ray-shuffle-optimization:

> We have also created a custom Dask DataFrame optimizer that leverages Ray's ability to

Reviewer: ```suggestion
Dask on Ray provides a Dask DataFrame optimizer that leverages Ray's ability to
```

execute multiple-return tasks in order to speed up shuffling by as much as 4x on Ray.
Simply set the `dataframe_optimize` configuration option to our optimizer function, similar to how you specify the Dask-on-Ray scheduler:

.. code-block:: python

   import ray
   from ray.util.dask import ray_dask_get, dataframe_optimize
   import dask.dataframe as dd
   import numpy as np
   import pandas as pd
   import time

   # Start Ray.
   # Tip: If you're connecting to an existing cluster, use ray.init(address="auto").
   ray.init()

   # Set the scheduler to ray_dask_get, and set the Dask DataFrame optimizer to our
   # custom optimization function, this time using the config setter as a context manager.
   with dask.config.set(scheduler=ray_dask_get, dataframe_optimize=dataframe_optimize):
       npartitions = 100
       df = dd.from_pandas(pd.DataFrame(np.random.randint(0, 100, size=(10000, 2)), columns=["age", "grade"]), npartitions=npartitions)

>        # We set max_branch to infinity in order to ensure that the task-based shuffle

Reviewer: hard to read -- can we do 80 col width?
